common4j/src/main/com/microsoft/identity/common/java/util/QueryParamsAdapter.java

```diff
@@ -0,0 +1,201 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.util;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.TypeAdapter;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+import com.microsoft.identity.common.java.exception.ClientException;
+import com.microsoft.identity.common.java.logging.Logger;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import lombok.AllArgsConstructor;
+
+/**
+ * Class to serialize and deserialize query parameters from List<Map.Entry<String, String>> to json String
+ * and vice versa.
+ *
+ * NOTE: Even we no longer use Pair (Since it's android-only), we are keeping this the same
+ *       to maintain backcompat with serialized value from older common that still uses it.
+ */
+@AllArgsConstructor
+public class QueryParamsAdapter extends TypeAdapter<List<Map.Entry<String, String>>> {
+
+    private static final String TAG = QueryParamsAdapter.class.getSimpleName();
+
+    private static final Gson mGson;
+
+    /**
+     * If enabled, this will write a json string in the 'proper' format.
+     * i.e. {"eqp1":"1","eqp2","2"}.
+     *
+     * Otherwise, it will write in the backcompat 'List<Pair<String, String>' format.
+     * i.e. [{"first":"eqp1","second":"1"},{"first":"eqp2","second":"2"}]
+     * */
+    boolean mWriteProperFormat = false;
+
+    static {
+        final GsonBuilder gsonBuilder = new GsonBuilder();
+        gsonBuilder.registerTypeAdapter(
+                getListType(),
+                // Turn off the "proper" format for now, because it'll break backcompat.
+                new QueryParamsAdapter(false)
+        );
```

Comment:
I'm wondering if this entire thing needs to live in some other class that bootstraps gson with custom adapters? Are we already doing that somewhere? If this is is the only instance then I guess it is fine to leave it here.

Reply:
Right now we're using the static class to serialize (and get ClientException - which i just added), so this will be the only place.